Library/OcCryptoLib/RsaDigitalSign.c

```diff
@@ -0,0 +1,511 @@
+/** @file
+
+OcCryptoLib
+
+Copyright (c) 2018, savvas
+
+All rights reserved.
+
+This program and the accompanying materials
+are licensed and made available under the terms and conditions of the BSD License
+which accompanies this distribution.  The full text of the license may be found at
+http://opensource.org/licenses/bsd-license.php
+
+THE PROGRAM IS DISTRIBUTED UNDER THE BSD LICENSE ON AN "AS IS" BASIS,
+WITHOUT WARRANTIES OR REPRESENTATIONS OF ANY KIND, EITHER EXPRESS OR IMPLIED.
+
+**/
+
+/**
+  Copyright (c) 2014 The Chromium OS Authors. All rights reserved.
+  Use of this source code is governed by a BSD-style license that can be
+  found in the LICENSE file.
+
+  Implementation of RSA signature verification which uses a pre-processed key
+  for computation.
+**/
+
+#ifdef EFIAPI
+#include <Library/BaseMemoryLib.h>
+#endif
+
+#include <Library/DebugLib.h>
+#include <Library/BaseLib.h>
+#include <Library/OcGuardLib.h>
+#include <Library/OcCryptoLib.h>
+
+/**
+  PKCS#1 padding (from the RSA PKCS#1 v2.1 standard)
+
+  The DER-encoded padding is defined as follows :
+  0x00 || 0x01 || PS || 0x00 || T
+
+  T: DER Encoded DigestInfo value which depends on the hash function used,
+  for SHA-256:
+  (0x)30 31 30 0d 06 09 60 86 48 01 65 03 04 02 01 05 00 04 20 || H.
+
+  Length(T) = 51 octets for SHA-256
+
+  PS: octet string consisting of {Length(RSA Key) - Length(T) - 3} 0xFF
+ **/
+#define PKCS_PAD_SIZE (CONFIG_RSA_KEY_SIZE - SHA256_DIGEST_SIZE)
+
+STATIC  UINT8 mSha256Tail[] = {
+  0x00, 0x30, 0x31, 0x30, 0x0d, 0x06, 0x09, 0x60,
+  0x86, 0x48, 0x01, 0x65, 0x03, 0x04, 0x02, 0x01,
+  0x05, 0x00, 0x04, 0x20
+};
+
+UINT64
+Mula32 (
+  UINT32  A,
+  UINT32  B,
+  UINT32  C
+  )
+{
+  UINT64 Ret = A;
+
+  Ret *= B;
+  Ret += C;
+  return Ret;
+}
+
+UINT64
+Mulaa32 (
+  UINT32  A,
+  UINT32  B,
+  UINT32  C,
+  UINT32  D
+  )
+{
+  UINT64 Ret = A;
+
+  Ret *= B;
+  Ret += C;
+  Ret += D;
+  return Ret;
+}
+
+//
+//  A[] -= Mod
+//
+STATIC
+VOID
+SubMod (
+  RSA_PUBLIC_KEY  *Key,
+  UINT32          *A
+  )
+{
+  INT64  B     = 0;
```

Comment:
Does not this potentially induce undefined behaviour due to being signed during the addition? This looks very worrying to me.

Reply:
It seems safe to me to change B to (UINT64) in this context on the compilers we support. Please do that when the code compiles though, so that we do not introduce more issues prior to validation step though.

0 assignments are also either not needed (for Index) or better to be moved below (for B).